### ISSUE TITLE
Virtual id setter called twice during model create

### ISSUE THREAD

**Issue by sdunnin** [typegoose/typegoose#403]
I have a parent Typegoose model class that maps an `id` virtual property to the mongo `_id` property and then uses the transform schema option to remove `_id` from result exposed to the application.

This all works quite well except for a certain scenario where we seed the database with static data having pre-defined id values.

**Parent Model Class**
```ts
import { mongoose, prop } from '@typegoose/typegoose';

import { RestResource } from '@nevco-one/shared';

export class ResourceDbModel implements RestResource {
  public get id() {
    return this._id.toString();
  }
  public set id(val: string) {
    if (!this._id) {
      this._id = new mongoose.Types.ObjectId(val);
    }
  }
  @prop()
  _id!: mongoose.Types.ObjectId;

  @prop()
  ver!: number;
}
```

**Schema Options (Applied to inherited classes)**
```ts
export const schemaOptions = {
  toJSON: {
    virtuals: true,
    versionKey: false,
    transform: (_doc: any, ret: { _id: any }, _opt: any) => {
      delete ret._id;
      return ret;
    }
  },
  toObject: {
    virtuals: true,
    versionKey: false,
    transform: (_doc: any, ret: { _id: any }, _opt: any) => {
      delete ret._id;
      return ret;
    }
  }
};
```

What I am seeing is that the virtual setter is being called twice. First time with the application supplied `id` value and second time with what appears to be a newly generated id by Mongo.  So what worked for me was to check to see if `_id` already had a value `if (!this._id)` and only set if it did not.

I was wondering if someone might have insight to the way Typegoose getters and setters work with Mongoose to maybe help explain why this is happening?

## Versions
-  typegoose: 7.4.1
 - mongoose: 5.10.7

**Comment by hasezoey** [comment]
is this only happening on extending classes, or on `ResourceDbModel` itself too (or cant test)?

**Comment by sdunnin** [comment]
ResourceDbModel should probably be abstract because it is never used as a model itself.

**Comment by sdunnin** [comment]
With the above configuration I then ran into an issue where Mongoose create(model) would fail with the error 
`MongooseError: document must have an _id before saving`. This presumably was caused by the `@prop()` decoration on the _id field causing it to expect one to be provided by the application.

My next attempt was to add `{ default: new mongoose.Types.ObjectId() }` to the _id's prop. Problem with that was it applied the same id to every document when an array of docs was passed into create causing a duplicate key error.

The solution that I found to work in both cases (id supplied in seed data, and normal use case where application lets Db supply the id) was to make the following changes to my initial configuration.

In `schemaOptions` I added `_id: false` so that Mongo expects the application to always provide the id for new documents. 

Then in `ResourceDbModel` in the prop for _id I made the default option a function returning a new ObjectId
```
  @prop({
    default: () => {
      return new mongoose.Types.ObjectId();
    }
  })
  _id!: mongoose.Types.ObjectId;
```
This has worked well so far to support both document creation scenarios.
